### DISCUSSION TITLE
IDE's & working development environments

### DISCUSSION THREAD

**Discussion by ischoegl** [Cantera/enhancements#96] [General]
There presumably is an entire range of solutions on this topic, but I wanted to start a discussion to see examples for how various community members have set up their tool-chains for *development* purposes (i.e. contributing to the Cantera repository, not just using the package). This certainly depends on individual preferences, but I believe it would be interesting to see what works, especially as related questions occasionally come up on the user group.

If you have a description of an environment that you'd like to share, please post a new comment below.

**Comment by ischoegl** [comment]
**Bare-bones in docker environment.** ... based on one of my own [UG responses](https://groups.google.com/g/cantera-users/c/Shy1QXNDhV0/m/vqItkCPaAgAJ) (which may or may not have provided meaningful feedback), this is (still) the description of a working process: 

> I compile my own cantera fork within a docker container, and use the equivalent of this approach: https://rubberduckdebugging.com/, albeit without the duck. Jests aside, I mainly work through the python interface and tightly couple C++ code with corresponding cython wrappers, i.e. I can interactively probe behavior from ipython. I have used valgrind as a big gun if there's hard to track-down memory errors. Python code is easily developed within jupyter (inheritance!). I am aware that there are more sophisticated approaches, but rarely found a need for them.

I originally chose docker as it allowed me to keep the development environment completely separate from my operating system (the setup - essentially an enhanced Ubuntu 18.04 container - is [uploaded here](https://github.com/ischoegl/cantera-docker)). To date, I mostly used emacs for editing, although I have started to use VS code for other projects. My OS is Fedora, but this is secondary for this setup. As mentioned above, the approach itself is not very sophisticated, so I do not expect this to end up as being a recommended solution.

To run quick tests on MATLAB, I have a second Win10 tool-chain based on Visual Studio Community ([posted here as a gist](https://gist.github.com/ischoegl/ab777f871a48d7f40ecc32c1c849faa3)); while it presumably can be extended, I have not used it for development.

**Reply by bryanwweber** [reply]
Nice! Do you use any VS Code extensions to develop inside the Docker container?

I'm very interested in developing a YUM/DNF(?) package for RPM based distros. Would Fedora be a good choice to work on that? Or would CentOS/Rocky Linux be better do you think?

**Reply by ischoegl** [reply]
VS Code extensions ... Nope, but I’ve seen some descriptions (I.e. extensions can apparently be used) so that’s on the to-do list. I either need to get a real rubber duck or learn something new ...

Ad Fedora, it works great (I usually skip every other release and then wait a couple of weeks). I have no idea about packaging though ...

**Comment by bryanwweber** [comment]
Great idea!

I use a MacBook Pro. Until the last month, I had a 2016 Intel with macOS 10.14. I now use a 2021 Apple Silicon model with 11.2. I also remote to a Windows desktop in my on-campus office that I haven't been to in a year, although usually only to test Windows specific stuff.

For code editing, I use 95% VS Code. I started in Notepad++ some decade+ ago, and wrote at least one paper in LaTeX using npp. From there, I spent a few years in Sublime Text, before moving to Atom for the GitHub integration. I can't recall why I switched to VS Code, I think it was probably the superior extensions, especially for LaTeX editing.

The other 5% is Vim when I'm at a terminal. I forced myself to learn more than just `:wq` last summer when I was working on a very old laptop that couldn't run VS Code, and now I use it over Nano.

For compilers, I tend to just use the default Apple `clang`, but I'll occasionally test with GCC and LLVM `clang` installed by Homebrew. I figured out how to set breakpoints in C++ code using VS Code and trigger them from a Python script, I'll have to post that into a blog post...

I tend to keep a copy of Matlab around for testing.

For Python management, I use about 95% `pyenv` with the `pyenv-virtualenv` plugin. I love love love the feature of being able to set a virtual environment that's automatically activated when I enter a directory. The other 5% is conda, mostly for testing the Cantera recipe. I installed `mambaforge` on my new laptop, it's worked quite well. It's like miniconda except using the `mamba` solver and with `conda-forge` as the default channel. I used to use `conda` exclusively, but one day I went looking for auto-environment-activation and conda couldn't do it. I bet they can now though.

**Reply by ischoegl** [reply]
:+1: for `pyenv` - use this one also with `miniconda` as a backend ... it’s really nice indeed!
